source/gboardforensics/gatherers/translatecache.d

```diff
@@ -0,0 +1,124 @@
+module gboardforensics.gatherers.translatecache;
+
+import gboardforensics.gatherers : IGatherer;
+import gboardforensics.models : TranslateCache;
+
+import std.algorithm : filter, findSplitAfter, findSplitBefore, map, sort, splitter, until;
+import std.file : dirEntries, DirEntry, read, readText, SpanMode;
+import std.path : baseName, buildPath, extension, stripExtension;
+import std.range : assocArray, back, chunks, front, split, tail;
+import std.regex : matchFirst, regex;
+import std.stdio : File;
+import std.typecons : No, tuple;
+
+class TranslateCacheGatherer : IGatherer
+{
+	this(DirEntry _dir)
+		in (_dir.isDir())
+	{
+		this._dir = _dir;
+	}
+
+	void gather()
+	{
+		import std.array : array;
+
+		_cache.data = _dir.name
+			.dirEntries(SpanMode.shallow)
+			.filter!"a.isFile"
+			.filter!(n => n.baseName.matchFirst(regex(r".+\.[0|1]$")))
+			.array
+			.sort
+			.chunks(2)
+			.map!array
+			.filter!(c => c.front.baseName.stripExtension() == c.back.baseName.stripExtension())
+			.filter!(c => c.front.extension() != c.back.extension())
+			.map!(c => TranslateCacheFiles(File(c.front), File(c.back)))
+			.map!toData
+			.array;
```

Comment:
It is fine to do this way, although, you can do this more efficiently with an associative array.

https://run.dlang.io/is/FNaUim

Reply:
Yup, this is actually really clever! I didn't know fold accepted a seed as the starter argument! You can go even further with this and make the call to the AA lazy.